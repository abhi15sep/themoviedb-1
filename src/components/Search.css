.Search-main {
    margin: 2em auto;
    position: relative;
}

.Search-clear {
    position: absolute;
    top: 1.2em;
    right: 1em;
    z-index: 3;
    cursor: pointer;
}
