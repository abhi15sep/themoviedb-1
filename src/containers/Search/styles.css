.Search-main {
    position: relative;
}

.Search-clear {
    position: absolute;
    top: .5em;
    right: 1em;
    z-index: 3;
    cursor: pointer;
}
