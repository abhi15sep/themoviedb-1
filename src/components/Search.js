import React, { Component } from 'react';
import { Grid, Row, Col, FormGroup, FormControl, Glyphicon } from 'react-bootstrap';
import SearchResults from './SearchResults';
import api from '../api/';
import './Search.css';

const MIN_LENGTH_SEARCH = 3;

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: false,
      error: false,
      query: '',
      movies: [],
    };

    this.handleChangeInputSearch = this.handleChangeInputSearch.bind(this);
    this.handleResetSearch = this.handleResetSearch.bind(this);
  }

  componentDidMount() {
    this.isAlreadyMounted = true;
  }

  componentWillUnmount() {
    this.isAlreadyMounted = false;
  }

  handleChangeInputSearch(event) {
    const query = event.target.value;
    this.setState({ query });
    this.checkMinLength(query);
  }

  handleResetSearch() {
    this.setState({
      query: '',
      movies: [],
    });
  }

  checkMinLength(query) {
    const { length } = query;

    if (length < MIN_LENGTH_SEARCH) {
      this.setState({
        movies: [],
      });
    }

    if (length >= MIN_LENGTH_SEARCH) {
      this.searchMovie(query);
    }
  }

  searchMovie(query) {
    api.search(query)
      .then((data) => {
        if (this.isAlreadyMounted) {
          this.setState({
            movies: data.results,
            loading: false,
          });
        }
      })
      .catch(() => {
        if (this.isAlreadyMounted) {
          this.setState({
            error: true,
          });
        }
      });
  }

  renderCloseIcon() {
    const { query } = this.state;

    if (query.length >= MIN_LENGTH_SEARCH) {
      return (
        <Glyphicon
          glyph="remove"
          className="Search-clear"
          onClick={this.handleResetSearch}
        />
      );
    }

    return null;
  }

  render() {
    const {
      query,
      error,
      loading,
      movies,
    } = this.state;

    return (
      <Grid className="Search-main">
        <Row>
          <Col xs={12}>
            <form className="Search-form form-inline">
              <div className="input-group input-group-lg">
                <FormGroup bsSize="large">
                  <FormControl
                    type="text"
                    value={query}
                    onChange={this.handleChangeInputSearch}
                    placeholder="Search a movie"
                    className="Search-input form-control"
                  />
                </FormGroup>
                {this.renderCloseIcon()}
                <SearchResults
                  error={error}
                  movies={movies}
                  loading={loading}
                  onMovieSelected={this.handleResetSearch}
                />
              </div>
            </form>
          </Col>
        </Row>
      </Grid>
    );
  }
}

export default Search;
