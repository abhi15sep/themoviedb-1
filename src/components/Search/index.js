import React from 'react';
import { Row, Col } from 'react-bootstrap';
import SearchForm from '../SearchForm';
import SearchClear from '../SearchClear';
import SearchResults from '../SearchResults';
import './styles.css';

/**
 * Component for display the search of the app
 *
 * @returns {Search} - The react component
 */
export default function Search() {
  return (
    <div className="Search-main navbar-text navbar-right">
      <Row>
        <Col xs={12}>
          <form className="Search-form form-inline">
            <div className="input-group input-group-lg">
              <SearchForm />
              <SearchClear />
              <SearchResults />
            </div>
          </form>
        </Col>
      </Row>
    </div>
  );
}
