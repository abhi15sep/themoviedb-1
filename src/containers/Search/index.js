import { FaBan } from 'react-icons/fa';
import React from 'react';
import {
  Row,
  Col,
  FormGroup,
  FormControl,
} from 'react-bootstrap';
import api from '../../api/index';
import SearchResults from '../../components/SearchResults';
import './styles.css';

const MIN_LENGTH_SEARCH = 3;

/**
 * Container for display the search of the app
 *
 * @returns {Search} - The react component
 */
export default function Search() {
  const [loading, setLoading] = React.useState(false);
  const [error, setError] = React.useState(false);
  const [query, setQuery] = React.useState('');
  const [movies, setMovies] = React.useState([]);

  const handleChangeInputSearch = (event) => {
    const { value } = event.target;
    setQuery(value);
  };

  const handleResetSearch = () => {
    setQuery('');
    setMovies([]);
  };

  React.useEffect(() => {
    const { length } = query;

    if (length >= MIN_LENGTH_SEARCH) {
      api.search(query)
        .then((data) => {
          const { results } = data;
          setMovies(results);
          setLoading(false);
        })
        .catch(() => {
          setError(true);
          setLoading(false);
        });
    }
  }, [query]);

  const renderCloseIcon = () => {
    if (query.length >= MIN_LENGTH_SEARCH) {
      return (
        <div
          tabIndex="0"
          role="button"
          className="Search-clear"
          onClick={handleResetSearch}
          onKeyPress={handleResetSearch}
        >
          <FaBan />
        </div>
      );
    }

    return null;
  };

  return (
    <div className="Search-main navbar-text navbar-right">
      <Row>
        <Col xs={12}>
          <form className="Search-form form-inline">
            <div className="input-group input-group-lg">
              <FormGroup>
                <FormControl
                  size="lg"
                  type="text"
                  value={query}
                  onChange={handleChangeInputSearch}
                  placeholder="Search a movie"
                  className="Search-input form-control"
                />
              </FormGroup>
              {renderCloseIcon()}
              <SearchResults
                error={error}
                movies={movies}
                loading={loading}
                onMovieSelected={handleResetSearch}
              />
            </div>
          </form>
        </Col>
      </Row>
    </div>
  );
}
