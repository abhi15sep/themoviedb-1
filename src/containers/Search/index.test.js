import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import Search from './index';

test('renders correctly', () => {
  const props = {
    history: {
      push: () => {},
    },
  };

  const component = renderer.create(<MemoryRouter><Search {...props} /></MemoryRouter>);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
