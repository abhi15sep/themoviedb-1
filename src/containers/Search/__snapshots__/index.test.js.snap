// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Search-main container"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <form
        className="Search-form form-inline"
      >
        <div
          className="input-group input-group-lg"
        >
          <div
            className="form-group form-group-lg"
          >
            <input
              className="Search-input form-control form-control"
              id={undefined}
              onChange={[Function]}
              placeholder="Search a movie"
              type="text"
              value=""
            />
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
